app/services/post_alerter.rb

```diff
@@ -268,19 +268,31 @@ def category_watchers(topic)
   end
 
   def category_or_tag_muters(topic)
+    user_option_sql_fragment =
+      if SiteSetting.watched_precedence_over_muted
+        <<~SQL
+        INTERSECT
+        SELECT user_id FROM user_options WHERE user_options.watched_precedence_over_muted IS false
+        SQL
+      else
+        <<~SQL
+        EXCEPT
+        SELECT user_id FROM user_options WHERE user_options.watched_precedence_over_muted IS true
+        SQL
+      end
+
     user_ids_sql = <<~SQL
-      SELECT user_id FROM category_users WHERE category_id = #{topic.category_id.to_i} AND notification_level = #{CategoryUser.notification_levels[:muted]}
-      UNION
-      SELECT user_id FROM tag_users tu JOIN topic_tags tt ON tt.tag_id = tu.tag_id AND tt.topic_id = #{topic.id} AND tu.notification_level = #{TagUser.notification_levels[:muted]}
-      EXCEPT
-      SELECT user_id FROM topic_users tus WHERE tus.topic_id = #{topic.id} AND tus.notification_level = #{TopicUser.notification_levels[:watching]}
-    SQL
-    User
-      .where("id IN (#{user_ids_sql})")
-      .joins("LEFT JOIN user_options ON user_options.user_id = users.id")
-      .where(
-        "user_options.watched_precedence_over_muted IS false OR (user_options.watched_precedence_over_muted IS NULL AND #{!SiteSetting.watched_precedence_over_muted})",
+      (
+        SELECT user_id FROM category_users WHERE category_id = #{topic.category_id.to_i} AND notification_level = #{CategoryUser.notification_levels[:muted]}
+        UNION
+        SELECT user_id FROM tag_users tu JOIN topic_tags tt ON tt.tag_id = tu.tag_id AND tt.topic_id = #{topic.id} WHERE tu.notification_level = #{TagUser.notification_levels[:muted]}
+        EXCEPT
+        SELECT user_id FROM topic_users tus WHERE tus.topic_id = #{topic.id} AND tus.notification_level = #{TopicUser.notification_levels[:watching]}
       )
+      #{user_option_sql_fragment}
```

Comment:
Just an idea, you might want to select this as a CTE and then join against `category_users`, `tag_users` and `topic_users` to reduce the number of row scans for those queries.